charms/kserve-controller/src/charm.py

```diff
@@ -362,105 +381,50 @@ def _generate_gateways_context(self) -> dict:
+            try:
+                getattr(self._stored, cert_attribute)
+            except AttributeError:
+                self._gen_certs()
+                break
+        log.info("Certificates already exist.")
+
+    def _gen_certs(self):
+        """Refresh the certificates, overwriting them if they already existed."""
+        log.info("Generating certificates..")
+        certs = gen_certs(
+            service_name=self._service_name,
+            namespace=self._namespace,
+            webhook_service=self._webhook_service,
         )
+        for k, v in certs.items():
+            setattr(self._stored, k, v)
+
+    def _upload_certs_to_container(
+        self, container: Container, destination_path: str, certs_store: StoredState
+    ) -> None:
+        """Upload generated certs to container.
+
+        Args:
+            container (Container): the container object to push certs to.
+            destination_path (str): path in str format where certificates will
+                be stored in the container.
+            certs_store (StoredState): an object where the certificate contents are stored.
+        """
+        try:
+            self._check_container_connection(container)
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
 
-        check_call(["openssl", "genrsa", "-out", "/run/ca.key", "2048"])
-        check_call(["openssl", "genrsa", "-out", "/run/server.key", "2048"])
-        check_call(
-            [
-                "openssl",
-                "req",
-                "-x509",
-                "-new",
-                "-sha256",
-                "-nodes",
-                "-days",
-                "3650",
-                "-key",
-                "/run/ca.key",
-                "-subj",
-                "/CN=127.0.0.1",
-                "-out",
-                "/run/ca.crt",
-            ]
-        )
-        check_call(
-            [
-                "openssl",
-                "req",
-                "-new",
-                "-sha256",
-                "-key",
-                "/run/server.key",
-                "-out",
-                "/run/server.csr",
-                "-config",
-                "/run/ssl.conf",
-            ]
-        )
-        check_call(
-            [
-                "openssl",
-                "x509",
-                "-req",
-                "-sha256",
-                "-in",
-                "/run/server.csr",
-                "-CA",
-                "/run/ca.crt",
-                "-CAkey",
-                "/run/ca.key",
-                "-CAcreateserial",
-                "-out",
-                "/run/cert.pem",
-                "-days",
-                "365",
-                "-extensions",
-                "v3_ext",
-                "-extfile",
-                "/run/ssl.conf",
-            ]
-        )
+        self.container.push(f"{destination_path}/tls.key", certs_store.key, make_dirs=True)
```

Comment:
Should we add error handling here in order to log that we cannot push?

Reply:
good question.  What do we want it to do here if it cannot upload?  Should it:
1. log and ignore (eg: keep running the charm)
2. log and Block
3. (what it does now) go into error
?

If we do (2), I'm not sure what we'd tell the user to do about it.  With (3), at least the charm keeps retrying because the hook failed.